File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/infrastructure/PostgresValorantSkinRepository.java
package fr.huiitre.tools.modules.riot.valorant.infrastructure;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.sql.Timestamp;
import java.time.LocalDateTime;

import org.springframework.jdbc.core.JdbcTemplate;

import fr.huiitre.tools.modules.riot.valorant.application.catalog.ports.ValorantSkinRepository;
import fr.huiitre.tools.modules.riot.valorant.application.skin.view.ValorantSkinLevelView;
import fr.huiitre.tools.modules.riot.valorant.application.skin.view.ValorantSkinView;

public class PostgresValorantSkinRepository implements ValorantSkinRepository {

    private final JdbcTemplate jdbcTemplate;

    public PostgresValorantSkinRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    private static final String SELECT_WITH_LEVELS = """
            SELECT s.id, s.asset_id, s.name, s.icon_url, s.tier_uuid, s.content_tier_uuid, s.weapon_id,
                   l.asset_id AS level_asset_id, l.level_index, l.display_icon_url, l.streamed_video_url,
                   (us.id IS NOT NULL) as owned, us.created_at as owned_at,
                   (w.id IS NOT NULL) as watched, w.created_at as watched_at
            FROM tools_riot.valorant_weapon_skins s
            LEFT JOIN tools_riot.valorant_skin_levels l ON l.skin_id = s.id
            LEFT JOIN tools_riot.valorant_user_skins us ON us.skin_id = s.id AND us.user_id = ?
            LEFT JOIN tools_riot.valorant_skin_watchlist w ON w.skin_id = s.id AND w.user_id = ?
            """;

    @Override
    public List<ValorantSkinView> findAll(Long userId) {
        final String sql = SELECT_WITH_LEVELS + " ORDER BY s.name, l.level_index";
        return buildMany(sql, userId, userId);
    }

    @Override
    public Optional<ValorantSkinView> findById(Long id, Long userId) {
        final String sql = SELECT_WITH_LEVELS + " WHERE s.id = ? ORDER BY l.level_index";
        return buildSingle(sql, userId, userId, id);
    }

    @Override
    public Optional<ValorantSkinView> findByAssetId(UUID assetId, Long userId) {
        final String sql = SELECT_WITH_LEVELS + " WHERE s.asset_id = ? ORDER BY l.level_index";
        return buildSingle(sql, userId, userId, assetId);
    }

    @Override
    public Optional<ValorantSkinView> findByLevelAssetId(UUID levelAssetId, Long userId) {
        final String sql = SELECT_WITH_LEVELS + """
                 WHERE s.id = (SELECT skin_id FROM tools_riot.valorant_skin_levels WHERE asset_id = ?)
                 ORDER BY l.level_index
                """;
        return buildSingle(sql, userId, userId, levelAssetId);
    }

    @Override
    public List<ValorantSkinView> findAllByWeaponId(Long weaponId, Long userId) {
        final String sql = SELECT_WITH_LEVELS + " WHERE s.weapon_id = ? ORDER BY s.name, l.level_index";
        return buildMany(sql, userId, userId, weaponId);
    }

    @Override
    public List<ValorantSkinView> findAllByTierUuid(UUID tierUuid, Long userId) {
        final String sql = SELECT_WITH_LEVELS + " WHERE s.tier_uuid = ? ORDER BY s.name, l.level_index";
        return buildMany(sql, userId, userId, tierUuid);
    }

    @Override
    public List<ValorantSkinView> findAllOwnedByUserId(Long userId) {
        final String sql = SELECT_WITH_LEVELS + " WHERE us.id IS NOT NULL ORDER BY s.name, l.level_index";
        return buildMany(sql, userId, userId);
    }

    @Override
    public List<ValorantSkinView> findAllWatchedByUserId(Long userId) {
        final String sql = SELECT_WITH_LEVELS + " WHERE w.id IS NOT NULL ORDER BY s.name, l.level_index";
        return buildMany(sql, userId, userId);
    }

    private List<ValorantSkinView> buildMany(String sql, Object... params) {
        return jdbcTemplate.query(sql, rs -> {
            LinkedHashMap<Long, ValorantSkinView> map = new LinkedHashMap<>();
            List<ValorantSkinLevelView> levels = new ArrayList<>();
            long lastSkinId = -1;

            while (rs.next()) {
                long skinId = rs.getLong("id");

                if (skinId != lastSkinId) {
                    if (lastSkinId != -1) {
                        rebuildLast(map, lastSkinId, levels);
                        levels = new ArrayList<>();
                    }
                    map.put(skinId, new ValorantSkinView(
                            skinId,
                            rs.getObject("asset_id", UUID.class),
                            rs.getString("name"),
                            rs.getString("icon_url"),
                            rs.getObject("tier_uuid", UUID.class),
                            rs.getObject("content_tier_uuid", UUID.class),
                            rs.getObject("weapon_id", Long.class),
                            new ArrayList<>(),
                            rs.getBoolean("owned"),
                            rs.getBoolean("watched"),
                            toLocalDateTime(rs.getTimestamp("owned_at")),
                            toLocalDateTime(rs.getTimestamp("watched_at"))));
                    lastSkinId = skinId;
                }

                String levelAssetIdStr = rs.getString("level_asset_id");
                if (levelAssetIdStr != null) {
                    levels.add(new ValorantSkinLevelView(
                            UUID.fromString(levelAssetIdStr),
                            rs.getInt("level_index"),
                            rs.getString("display_icon_url"),
                            rs.getString("streamed_video_url")));
                }
            }

            if (lastSkinId != -1) {
                rebuildLast(map, lastSkinId, levels);
            }

            return new ArrayList<>(map.values());
        }, params);
    }

    private Optional<ValorantSkinView> buildSingle(String sql, Object... params) {
        return jdbcTemplate.query(sql, rs -> {
            ValorantSkinView skin = null;
            List<ValorantSkinLevelView> levels = new ArrayList<>();

            while (rs.next()) {
                if (skin == null) {
                    skin = new ValorantSkinView(
                            rs.getLong("id"),
                            rs.getObject("asset_id", UUID.class),
                            rs.getString("name"),
                            rs.getString("icon_url"),
                            rs.getObject("tier_uuid", UUID.class),
                            rs.getObject("content_tier_uuid", UUID.class),
                            rs.getObject("weapon_id", Long.class),
                            levels,
                            rs.getBoolean("owned"),
                            rs.getBoolean("watched"),
                            toLocalDateTime(rs.getTimestamp("owned_at")),
                            toLocalDateTime(rs.getTimestamp("watched_at")));
                }

                String levelAssetIdStr = rs.getString("level_asset_id");
                if (levelAssetIdStr != null) {
                    levels.add(new ValorantSkinLevelView(
                            UUID.fromString(levelAssetIdStr),
                            rs.getInt("level_index"),
                            rs.getString("display_icon_url"),
                            rs.getString("streamed_video_url")));
                }
            }

            return Optional.ofNullable(skin);
        }, params);
    }

    private void rebuildLast(LinkedHashMap<Long, ValorantSkinView> map, long skinId, List<ValorantSkinLevelView> levels) {
        ValorantSkinView prev = map.get(skinId);
        map.put(skinId, new ValorantSkinView(
                prev.id(), prev.assetId(), prev.name(), prev.iconUrl(),
                prev.tierUuid(), prev.contentTierUuid(), prev.weaponId(), List.copyOf(levels),
                prev.owned(), prev.watched(), prev.ownedAt(), prev.watchedAt()));
    }

    private LocalDateTime toLocalDateTime(Timestamp ts) {
        return ts != null ? ts.toLocalDateTime() : null;
    }
}

File: api/Modules/Riot/Valorant/Infrastructure/PostgresValorantSkinRepository.cs
using Tools.Api.Modules.Riot.Common.Infrastructure;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Views;
using Tools.Api.Modules.Riot.Valorant.Application.Skin.Views;

namespace Tools.Api.Modules.Riot.Valorant.Infrastructure;

// Un skin porte ses niveaux, ses chromas et sa rareté. Les niveaux arrivent par jointure — une
// ligne par niveau, regroupée ici — et les chromas par une seconde requête sur les identifiants
// déjà connus : les joindre tous les deux d'un coup multiplierait les lignes entre eux.
public sealed class PostgresValorantSkinRepository(RiotDatabase database) : IValorantSkinRepository
{
    // accountId est comparé dans les jointures : nul, elles ne ramènent rien et les deux
    // drapeaux tombent à faux, sans exclure le skin.
    private const string SelectWithLevels = """
        SELECT s.id AS Id, s.asset_id AS AssetId, s.name AS Name, s.icon_url AS IconUrl,
               s.tier_uuid AS TierUuid, s.weapon_id AS WeaponId,
               (us.id IS NOT NULL) AS Owned, us.created_at AS OwnedAt,
               (w.id IS NOT NULL) AS Watched, w.created_at AS WatchedAt,
               l.asset_id AS LevelAssetId, l.level_index AS LevelIndex, l.name AS LevelName,
               l.level_item AS LevelItem, l.display_icon_url AS LevelDisplayIconUrl,
               l.streamed_video_url AS LevelStreamedVideoUrl,
               ct.id AS CtId, ct.asset_id AS CtAssetId, ct.name AS CtName, ct.dev_name AS CtDevName,
               ct.rank AS CtRank, ct.juice_value AS CtJuiceValue, ct.juice_cost AS CtJuiceCost,
               ct.highlight_color AS CtHighlightColor, ct.display_icon_url AS CtDisplayIconUrl
        FROM tools_riot.valorant_weapon_skins s
        LEFT JOIN tools_riot.valorant_skin_levels l ON l.skin_id = s.id
        LEFT JOIN tools_riot.valorant_user_skins us ON us.skin_id = s.id AND us.valorant_account_id = @AccountId
        LEFT JOIN tools_riot.valorant_skin_watchlist w ON w.skin_id = s.id AND w.valorant_account_id = @AccountId
        LEFT JOIN tools_riot.valorant_content_tiers ct ON ct.asset_id = s.content_tier_uuid
        """;

    private sealed record SkinRow(
        long Id,
        Guid AssetId,
        string Name,
        string? IconUrl,
        Guid? TierUuid,
        long? WeaponId,
        bool Owned,
        DateTime? OwnedAt,
        bool Watched,
        DateTime? WatchedAt,
        Guid? LevelAssetId,
        int? LevelIndex,
        string? LevelName,
        string? LevelItem,
        string? LevelDisplayIconUrl,
        string? LevelStreamedVideoUrl,
        long? CtId,
        Guid? CtAssetId,
        string? CtName,
        string? CtDevName,
        int? CtRank,
        int? CtJuiceValue,
        int? CtJuiceCost,
        string? CtHighlightColor,
        string? CtDisplayIconUrl);

    private sealed record ChromaRow(
        long SkinId,
        Guid AssetId,
        int ChromaIndex,
        string Name,
        string? DisplayIconUrl,
        string? FullRenderUrl,
        string? SwatchUrl,
        string? StreamedVideoUrl);

    public Task<List<ValorantSkinView>> FindAll(long? accountId) =>
        FindMany($"{SelectWithLevels} ORDER BY s.name, l.level_index", new { AccountId = accountId });

    public Task<ValorantSkinView?> FindById(long id, long? accountId) =>
        FindOne($"{SelectWithLevels} WHERE s.id = @Id ORDER BY l.level_index", new { AccountId = accountId, Id = id });

    public Task<ValorantSkinView?> FindByAssetId(Guid assetId, long? accountId) =>
        FindOne($"{SelectWithLevels} WHERE s.asset_id = @AssetId ORDER BY l.level_index", new { AccountId = accountId, AssetId = assetId });

    // Riot ne connaît que les UUID de niveaux : on remonte au skin parent avant de tout charger.
    public Task<ValorantSkinView?> FindByLevelAssetId(Guid levelAssetId, long? accountId) =>
        FindOne(
            $"""
             {SelectWithLevels}
             WHERE s.id = (SELECT skin_id FROM tools_riot.valorant_skin_levels WHERE asset_id = @LevelAssetId)
             ORDER BY l.level_index
             """,
            new { AccountId = accountId, LevelAssetId = levelAssetId });

    public Task<List<ValorantSkinView>> FindAllByWeaponId(long weaponId, long? accountId) =>
        FindMany($"{SelectWithLevels} WHERE s.weapon_id = @WeaponId ORDER BY s.name, l.level_index",
            new { AccountId = accountId, WeaponId = weaponId });

    public Task<List<ValorantSkinView>> FindAllByTierUuid(Guid tierUuid, long? accountId) =>
        FindMany($"{SelectWithLevels} WHERE s.tier_uuid = @TierUuid ORDER BY s.name, l.level_index",
            new { AccountId = accountId, TierUuid = tierUuid });

    public Task<List<ValorantSkinView>> FindAllOwnedByAccountId(long accountId) =>
        FindMany($"{SelectWithLevels} WHERE us.id IS NOT NULL ORDER BY s.name, l.level_index",
            new { AccountId = accountId });

    public Task<List<ValorantSkinView>> FindAllWatchedByAccountId(long accountId) =>
        FindMany($"{SelectWithLevels} WHERE w.id IS NOT NULL ORDER BY s.name, l.level_index",
            new { AccountId = accountId });

    private async Task<ValorantSkinView?> FindOne(string sql, object parameters) =>
        (await FindMany(sql, parameters)).FirstOrDefault();

    private async Task<List<ValorantSkinView>> FindMany(string sql, object parameters)
    {
        var rows = await database.Query<SkinRow>(sql, parameters);

        if (rows.Count == 0)
        {
            return [];
        }

        var chromasBySkinId = await FindChromas(rows.Select(row => row.Id).Distinct().ToArray());

        return rows
            .GroupBy(row => row.Id)
            .Select(group => ToView(group.First(), group, chromasBySkinId))
            .ToList();
    }

    private static ValorantSkinView ToView(
        SkinRow first,
        IEnumerable<SkinRow> rows,
        Dictionary<long, List<ValorantSkinChromaView>> chromasBySkinId)
    {
        // Un skin sans niveau produit tout de même une ligne, avec ses colonnes de niveau nulles.
        var levels = rows
            .Where(row => row.LevelAssetId is not null)
            .Select(row => new ValorantSkinLevelView(
                row.LevelAssetId!.Value,
                row.LevelIndex ?? 0,
                row.LevelName ?? string.Empty,
                row.LevelItem,
                row.LevelDisplayIconUrl,
                row.LevelStreamedVideoUrl))
            .ToList();

        return new ValorantSkinView(
            first.Id,
            first.AssetId,
            first.Name,
            first.IconUrl,
            first.TierUuid,
            ToContentTier(first),
            first.WeaponId,
            levels,
            chromasBySkinId.GetValueOrDefault(first.Id, []),
            first.Owned,
            first.Watched,
            first.OwnedAt,
            first.WatchedAt);
    }

    private static ValorantContentTierView? ToContentTier(SkinRow row) =>
        row.CtId is { } tierId && row.CtAssetId is { } tierAssetId
            ? new ValorantContentTierView(
                tierId,
                tierAssetId,
                row.CtName ?? string.Empty,
                row.CtDevName ?? string.Empty,
                row.CtRank ?? 0,
                row.CtJuiceValue ?? 0,
                row.CtJuiceCost ?? 0,
                row.CtHighlightColor,
                row.CtDisplayIconUrl)
            : null;

    private async Task<Dictionary<long, List<ValorantSkinChromaView>>> FindChromas(long[] skinIds)
    {
        var rows = await database.Query<ChromaRow>(
            """
            SELECT skin_id AS SkinId, asset_id AS AssetId, chroma_index AS ChromaIndex, name AS Name,
                   display_icon_url AS DisplayIconUrl, full_render_url AS FullRenderUrl,
                   swatch_url AS SwatchUrl, streamed_video_url AS StreamedVideoUrl
            FROM tools_riot.valorant_skin_chromas
            WHERE skin_id = ANY(@SkinIds)
            ORDER BY skin_id, chroma_index
            """,
            new { SkinIds = skinIds });

        return rows
            .GroupBy(row => row.SkinId)
            .ToDictionary(
                group => group.Key,
                group => group.Select(row => new ValorantSkinChromaView(
                    row.AssetId,
                    row.ChromaIndex,
                    row.Name,
                    row.DisplayIconUrl,
                    row.FullRenderUrl,
                    row.SwatchUrl,
                    row.StreamedVideoUrl)).ToList());
    }
}
